[FEATURE] Plugins API with Gallery Tab, Lora Multipliers UI As Official Plugins

Tophness opened the pull request:
You were talking about a plugins API.
I liked the idea, in fact I've wanted this for ages because it was one of the most powerful features of A1111, but thought it would take a long time and there would be a ton of issues hooking gradio I/O like I've already been dealing with, so I got started on it to see what that would look like. I was wrong. It's already done and actually makes working with gradio a little easier.

I have fully implemented my lora multipliers ui and gallery as default system plugins and moved all existing tabs to plugins.
PLUGINS.md details how you can make your own.
There's also now a plugin-manager plugin that adds a plugins tab that allows you to enable, disable, update, uninstall, reinstall and install plugins from github with the full setup routine including requirements.txt

So far it can create new tabs, read and write to existing components, insert new components after a named component and access wgp.py global functions and vars.
Plugins in the plugins folder are automatically discovered.

Todo (after this gets merged to main):
- Access and modify the queue (this can already be achieved by accessing the global functions, but an actual API would probably be better for devs to play around with it in future)
- Ability to add new models (again, can already be achieved, but a native API would be better)

<img width="2320" height="1323" alt="image" src="https://github.com/user-attachments/assets/9fc7fb26-c150-4f13-a0ad-2c4a8d130206" />

Gunther-Schulz commented:
Great stuff! I am going on vacation soon, but will have a look when I get back.

Tophness commented:
Oh fuck yeah I got it.
You can now insertAfter() a named gradio component.
Now I'll try make a plugin for my lora_multipliers_ui branch.

<img width="1135" height="551" alt="image" src="https://github.com/user-attachments/assets/907a1e1e-46a4-4eb5-9e07-031bd50dc9c0" />

Tophness commented:
Gallery tab done. That required me to add global functions and vars to the plugin system too. Good test case.

deepbeepmeep commented:
very nice, all of this will be great material for v9 ! I am almost done with the stuff I had in mind for 8.99 and I will start to go deeper into the different PR

by the way I just did a little change to loras multipliers syntax since now there is also a "|" that can be added to separate loras accelerators / non accelerators. You should just treat as Space char

jsr184 commented:
Would it be suitable for tasks like adding new models into the Wan2GP? I had in mind adding HuMo support but at this point it was much easier to make a standalone UI for it using gradio rather than hack through Wan2GP to integrate it.

Tophness commented:
I think for stuff like controlnet models, mask editor models and such that require custom UI and other things that you might want to work alongside certain models, it could make sense, but for just regular model support, DBM has already made that pretty modular by the looks of it.

deepbeepmeep commented:
So I am done with 8.99. @Tophness Please merge your latest stuff with the main branch and I will start playing with it

deepbeepmeep commented:
> Would it be suitable for tasks like adding new models into the Wan2GP? I had in mind adding HuMo support but at this point it was much easier to make a standalone UI for it using gradio rather than hack through Wan2GP to integrate it.

It all depends if you feel confident with working on any2video.py and model.py. The main advantage of WanGP is the ready to use preprocessing pipeline (for instance extraction of poses), post processing (upsampling, mmaudio, ...) and the UI that allows the user to easily switch between inputs text / image prompts, ...

I had a look at Humo and it didnt seem hard to implement but I had no time for it. The only tricky stuff was that it was switching guidance in the middle of the gen and it is not a simple guidance switch (the conditions sent to the model are very different).

deepbeepmeep commented:
Well done ! All of this open some very interesting possibilites. 

My comments:

- Plugins
We should distinguish System Plugins from  Optional plugins. It doesnt make sense to remove the Configuration or About Tabs. The Guide tab should always be there as it is light and frequently used (and we dont want to have to answer questions whose answers are in the guide...).
So the Original defaults Tabs (the 'wan2gp' plugins) shouldn't be Selectable Plugins, the user shouldnt be able to remove them / disable them.  Matanyone is technically a plugin but is indispensable. There could be an exception like Download which is not indispensable but in any case should still come preinstalled.
The Last 3 tabs should always be in the same order for visual consitence, for instance: Plugins, Guide, About
Becareful the position of the Matanyone is hardcoded in the code.
Happy in any case to keep the plugin internal architecture for the system plugins.
Ideallly we should be able to change the order of custom plugins (system plugins order like or Gui or About should be hardcoded) and disable a plugin without desinstalling it.
We should be able to select non installed Plugins that are referenced in the WanGP Github repo

There is one important thing that is still missing I hadn't time to do but is crucial for advanced plugins: Generation Lock
For the moment it is impossible to use matanyone or any other plugin that needs to use cuda while a generation is in progress (we can't have two cuda processing the time as this messes everything up). I have created for the Prompt Enhancer a Generation Lock / Pause system that suspends a Video Generation while the prompt ehancer is working and then it automatically resumes the gen after. We should give access to every plugins (including matanyone) to this lock system so that the user can use a plugin for instance that performs a Cuda preprocessing / postprocessing (upsampling for instance)that is not offered by WanGP whenever he wants (not to have to wait that the gen is finised).

We should have a per (non system) plugin requirements.txt file to keep the plugin very modular and probably later a config menu per plugin (which could be directly in the plugin which is up to debate)

By the way, have you noticed, it is possible to fill any field of the Video Gen from another tab (but you shouldnt directly write in gradio elements otherwise the change wont be seen by the server) by accessing the user inputs dictionary. This is what I use with matayone to update fields like the control video, control mask, etc.. I think for a lot plugins it will make sense to be able to access easily this dictionary (for instance for an external prompt enhancer) to modify the settings on the fly. We should make this dictionary easily accessible in plugin API.

- loras UI
Any way to merge after a split ?
ui wont update if I modifiy directly the input
Personally I am not going use it, I would appreciate if it could be disabled by default as the Gradio interface is gettting slower and slower
I think this plugin is different from the Plugins that are tabs maybe we should call it Add On, or consider it as a System Extension (like the prompt enhancer) that can be enabled from the config tab. Indeed its hooks are hardcoded in the Video Gen tab.

- Gallery:
Was very slow the first time I launched it (a few minutes), I have 4000 videos / images. Maybe you should record thumbnails somewhere to save time. Will it work on Linux ? If not this will be an issue or we will have two plugins catalogs.
I dont know if it related to the Gallery that is doing its work in the background but when I start WanGP everything is very slow, responsivness is bad. If that's the case the Gallery shouldn't start processing if it is not selected.
We should be able to send a Video or its start frame / end frame (specific frame has been asked many times but I am not sure one can select one with this gallery) to the Video Gen Gallery so that it can be used later in a new gen.

Tophness commented:
>       We should distinguish System Plugins from  Optional plugins. It doesnt make sense to remove the Configuration or About Tabs. The Guide tab should always be there as it is light and frequently used (and we dont want to have to answer questions whose answers are in the guide...).

I wouldn't mind disabling about or guides personally tbh. Once you've seen the about section once, it doesn't really serve any purpose anymore. Same with guides if you already know how to use it.

>       The Last 3 tabs should always be in the same order for visual consitence, for instance: Plugins, Guide, About

Can do. So far I thought you might say all that lol. I was mostly getting all the major functionality working first and set the orders kinda arbitrarily bc it's a single variable change when you define the new tab.

>       Becareful the position of the Matanyone is hardcoded in the code.

Yeah, had a lot of trouble getting the Matanyone tab to even show up cos of how that all worked. It is working right? Like you can actually use matanyone in the tab?


>       Ideallly we should be able to change the order of custom plugins (system plugins order like or Gui or About should be hardcoded) and disable a plugin without desinstalling it.

Yeah, I was planning on that too. The order isn't set by order of installation btw,
It's set by
        self.add_tab(
            label="Tab name",
            position=2
        )

Position is it's (optional) preferred position, and then the sorter will avoid conflicts if they both decide on the same position.
There's a setting to allow you different sort modes though, so I'll add a custom one for user sorting.


>       We should be able to select non installed Plugins that are referenced in the WanGP Github repo

No idea what you mean by this


> There is one important thing that is still missing I hadn't time to do but is crucial for advanced plugins: Generation Lock

Yeah I was planning that. Problem is I think to do that right might create merge conflicts, so I was leaving it until last.
The right way to do it imo would be to actually make wgp less naive of the plugin system and have generation lock be a less hardcoded function anyone can use, but right now it's just used for config and matanyone right?
It would probably help if you could do that too, I don't want to create some new generation lock system myself and have your old stuff still left lying around wasting code or causing conflicts.


> We should have a per (non system) plugin requirements.txt file to keep the plugin very modular

This feature already exists. Haven't tested it, but it's supposed to do that on install of a github plugin.

> and probably later a config menu per plugin (which could be directly in the plugin which is up to debate)

Agreed. You could already do this now by hooking globals and components, but a native API for this would be better.


> By the way, have you noticed, it is possible to fill any field of the Video Gen from another tab (but you shouldnt directly write in gradio elements otherwise the change wont be seen by the server) by accessing the user inputs dictionary. This is what I use with matayone to update fields like the control video, control mask, etc.. I think for a lot plugins it will make sense to be able to access easily this dictionary (for instance for an external prompt enhancer) to modify the settings on the fly. We should make this dictionary easily accessible in plugin API.

Not sure exactly what you mean. The gallery tab sends parameters to video gen when you join, and lora multipliers ui sets values on load.
Do you mean something that looks more like a specialized API just for that?
If so I agree, but it'd be good to get this merged soon to avoid merge conflicts and do that later.

>     * loras UI
>       Any way to merge after a split ?

Oh yeah, I was planning to do that and forgot.

>       ui wont update if I modifiy directly the input

Yeah, it did do that at one point, but I got rid of it while I was trying to debug stuff cos it makes the code way more complex.
Will add it back in eventually.

>       Personally I am not going use it, I would appreciate if it could be disabled by default as the Gradio interface is gettting slower and slower

Sure, but it does make everything much more easy and efficient.

>     * Gallery:
>       Was very slow the first time I launched it (a few minutes), I have 4000 videos / images. Maybe you should record thumbnails somewhere to save time.

Damn I thought I'd sped it up on windows. I have about that many too. A cache is a reasonable thing and I was thinking about it, I just have some concerns (See #)

>       Will it work on Linux ? If not this will be an issue or we will have two plugins catalogs.

It will work, the thumbnails will just have placeholders for now, but the rest of the functionality will be there, and they can see the preview frames of the video when they click on it anyway.

>       I dont know if it related to the Gallery that is doing its work in the background but when I start WanGP everything is very slow, responsivness is bad. If that's the case the Gallery shouldn't start processing if it is not selected.

Not the gallery, it only starts loading files when you go to the tab, and that's the only resource intensive process.

>       We should be able to send a Video or its start frame / end frame (specific frame has been asked many times but I am not sure one can select one with this gallery) to the Video Gen Gallery so that it can be used later in a new gen.

Not sure what you mean. You can sent start and end frames to the video generator. Are you saying send it from the gallery tab to the gallery tab within the video generator, and then be able to sent it from the video generator's gallery to the video generator?
Wouldn't that just be the same thing with extra steps?

Tophness commented:
Looks like the performance drop I was getting was actually introducing a new bug in the plugin system actually. It was too hard for me to find how I got the lora multiplier ui working on the edit tab cos the older stuff i had waiting to patch after a merge doesn't apply anymore, and the way I did it this time made a lot of duplicates of the routines to hook the edit tab through the plugin system.
I've removed that for now and i'll just try get it on the edit tab some other way, but in trying to track this down I cleaned up a lot of the js functions in the plugins and wgp.py.
Might be worth seeing if the gallery is still slow for you.
I did a lot of performance tracking and couldn't find anything that would explain the severity you were describing though.
lora multipliers ui ran for less than 0.0000s to do ui updates, gallery took 0.6362s to get thumbnails once and never again.
I then even tracked each function itself and it was a similar story. Only thing left was the js, and that does fit, because there were a lot of mutationobservers getting spawned, but that's harder to bench.

deepbeepmeep commented:
> > Yes, the Edit Tab for the Queue Editor is always visible
> 
> That's always been it's behaviour

What's the use for this ? it is very confusing. We dont know what we are editing if we arrive straight on it. We should see the Edit form only if we have actually selected a queue item before that. Once the edition is done it should completely disappear. 
I can understand that having the Edit Tab separate from the Main Video Gen tab allows the user to go back & forth and check the progress of its gens while editing one but it may be very risky if for instance the user delete the queue item being currently edited for instance. What happens when the currently edited item becomes the current one being processed ? do we cancel automatically the edition ? In that case I think we should pause the gen until the user presses save or cancel

To avoid conflicting actions it may be simpler to have the Edit mode replace the content of the main video gen tab (for the time of the edition) and restore it once the edition is done.

Tophness commented:
> What's the use for this ? it is very confusing. We dont know what we are editing if we arrive straight on it.
> We should see the Edit form only if we have actually selected a queue item before that. Once the edition is done it should completely disappear.

You could clear it if you try to switch between the generator and edit tabs, that would just sacrifice great convenience for avoiding confusion someone might get if the edit tab contains the item they were editing.
It should also show which item they were still editing at the top, so it shouldn't be a complete mystery.
You could put an extra note on there to just tell them they can continue editing the item where they left off.


> I can understand that having the Edit Tab separate from the Main Video Gen tab allows the user to go back & forth and check the progress of its gens while editing one but it may be very risky if for instance the user delete the queue item being currently edited for instance.

I haven't had the chance to test all these cases yet, although I have thought about how I would handle them if they're an issue.
If they edit a deleted item, that should be an easy 1 line conditional check. If it's null, don't do anything.

> What happens when the currently edited item becomes the current one being processed ? do we cancel automatically the edition ? In that case I think we should pause the gen until the user presses save or cancel

Pretty much the same as above, just don't try to do it. The item would be popped from queue, so it should be null and void anyway.
Could also notify them that it's generating or about to generate, then offer them a choice of: Pause current generation to edit it, Add the edition to queue as a new item or cancel.


> To avoid conflicting actions it may be simpler to have the Edit mode replace the content of the main video gen tab (for the time of the edition) and restore it once the edition is done.

I had it like this for my first rough draft that I was using for ages. It looks nice but in practice it was so much less efficient. A lot of times when you're generating long queues you're experimenting,
so it's helpful to have a consistent flow where you're changing 1 parameter at a time on the 1 tab and can see what you're doing, and having the editor jump in and override it with a whole bunch of different parameters makes you forget what you were doing and have to look through everything again

Tophness commented:
> bind events for plugins messes everything up: different events receive different object instances for the same user session, for instance the State object and especially the Gen key subdictionary are not the same between calls. Everything happens as if multiple user sessions were running at the sametime  for the same user. This prevents proper synchronization and breaks completely my lock system which needs to access the same gen dictionary. This will certainly break other things like the standard way of updating user inputs via a dictionary stored in State.
> In anycase you should find a way to keep the events in the Create_UI function, this makes it much harder to code (since the related gradio objects are in a different function) and the events appear as errors for Visual Code since the corresponding variables are not locals .

Just messaged you about this one on discord. It would probably require a bit of back and forth to make sure we understand each other since this sounds like a pretty significant change to the plugin system
